### DISCUSSION TITLE
Very slow live state update

### DISCUSSION THREAD

**Discussion by tkukushkin** [argoproj/argo-cd#6704] [Q&A]
Hello everyone!

We use argocd v2.0.4 and are experiencing a really slow kubernetes **live** state updates in ArgoCD UI. In other words, ArgoCD is very slowly updates information about kubernetes resources that has been updated or changed it states.

As an example, a rolling update with a new ReplicaSet initiated from ArgoCD is only shows up in a UI after 7-8 minutes it's been already up and running with all checks passed (as we can see in Kubernetes Dashboard / kubectl).

### Our setup:

App of apps, monorepo with raw kubernetes specs for about 400 argo applications and 25k kubernetes resources in total (from `argocd_cluster_api_resource_objects` metric).
All applications have autosync enabled:
```yaml
syncPolicy:
  automated:
    prune: true
    selfHeal: false
```

I would like to clarify right away that we do not have a problem with updating the **desired** state, our repo servers are fine with this monorepo, we believe that the problem with the application controller.

### Application controller settings:
```yaml
- argocd-application-controller
- --app-resync
- "3600"
- --status-processors
- "300"
- --operation-processors
- "75"
- --redis
- "argocd-redis:6379"
- --kubectl-parallelism-limit
- "150"
```

Kubernetes node, on which it is running, is utilized on 15-20% of CPU and 20% of memory.

Also we use dedicated Managed Redis in GCP for ArgoCD.

Can you tell us what we might be doing wrong and what we can do to fix this problem?

<details>
  <summary>Grafana dashboard screenshot</summary>

  ![argocd-grafana](https://user-images.githubusercontent.com/1482516/125430036-5f0040a8-d508-4a6e-9a42-9b3dd87d3d6b.png)

</details>

**Comment by swaragade** [comment]
Hi @tkukushkin,

Were you able to resolve this issue ? What is the solution ?

**Reply by tkukushkin** [reply]
No, we haven't found the solution for this problem. Also I'd like to mention, some time ago we started managing several kubernetes clusters by our Argo CD and we noticed, that this problem doesn't affect small clusters. So we have this problem with two big clusters and don't have it with 3 small clusters at the same time, while of the clusters are managed by one Argo CD.

So it is one of possible solutions, split workloads on several smaller clusters. It's not so easy, we wrote special service to sync `Endpoints` between clusters to achieve "multicluster" services. It watches all `Service` and `Endpoints` resources in all clusters and replicate `Endpoints` to other clusters (configured by annotation on `Service` resource). So our services can communicate with other services from other clusters as if they were in the same cluster. I hope sometime we will publish it to open source. Luckily Argo CD supports sharding of clusters. So you can scale argocd-application-controller and attach clusters to concrete instances of it. We have 3 instances of argocd-application-controller. 2 big clusters are attached to own instances, other 3 clusters are attached to one instance.

**Reply by swaragade** [reply]
Thanks for your inputs @tkukushkin, I will check in this direction
